agent/config/config.go

```diff
@@ -131,16 +129,129 @@ func (cfg *Config) Merge(rhs Config) *Config {
 	return cfg //make it chainable
 }
 
-// complete returns true if all fields of the config are populated / nonzero
-func (cfg *Config) complete() bool {
+// NewConfig returns a config struct created by merging environment variables,
+// a config file, and EC2 Metadata info.
+// The 'config' struct it returns can be used, even if an error is returned. An
+// error is returned, however, if the config is incomplete in some way that is
+// considered fatal.
+func NewConfig(ec2client ec2.EC2MetadataClient) (config *Config, err error) {
+	var errs []error
+	var errTmp error
+	envConfig, errTmp := environmentConfig() //Environment overrides all else
+	if errTmp != nil {
+		errs = append(errs, errTmp)
+	}
+	config = &envConfig
+	defer func() {
+		config.trimWhitespace()
+		config.Merge(DefaultConfig())
+		errTmp = config.validateAndOverrideBounds()
+		if errTmp != nil {
+			errs = append(errs, errTmp)
+		}
+		if len(errs) != 0 {
+			err = apierrors.NewMultiError(errs...)
+		} else {
+			err = nil
```

Comment:
I don't think you need this?

Reply:
why not?

Reply:
I thought the `err` is `nil` by default.

Reply:
Just noticed the named return here. urgh. i'll modify this to get rid of that.